File: quick_sort/quick_sort.c
// Quick Sort
#include <stdio.h>

#define SIZE 10

void swap(int *x, int *y) {
    int tmp = *x;
    *x = *y;
    *y = tmp;
}

int partition(int* arr, int start, int end) {

    int pivot = arr[end];
    int greaterElemIndex = start-1; // pointer for greater element
    int tmp;

    for (int i=start; i<end; i++) {
        if (arr[i] <= pivot) {
            greaterElemIndex++;
            swap(&arr[i], &arr[greaterElemIndex]);
        }
    }
    swap(&arr[end], &arr[greaterElemIndex+1]);
    return greaterElemIndex+1; // return the partition point
}

void quicksort(int* arr, int start, int end) {
    if(start < end) {
        int pivot = partition(arr, start, end);
        quicksort(arr, start, pivot-1);
        quicksort(arr, pivot+1, end);
    }
}

int main() {
    int arr[] = {4,10,8,7,6,2,3,12,14,5};

    quicksort(arr, 0, SIZE-1);

    for(int i = 0; i< SIZE; i++) {
        printf("%d ", arr[i]);
    }
    return 0;
}

File: quick_sort/quick_sort.py
# function to find the partition position
# return the index of the pivot
def partition(array, left, right):

    # choose the rightmost element as pivot (can be done differently)
    pivot = array[right]

    # pointer for greater element
    # view this pointer as the leftmost index of elements greater than pivot
    # so before returning, swap(array[greaterElemIndex], array[right]) to move pivot to its rightful position
    firstGreaterElemIndex = left

    # traverse through all elements and compare themw ith pivot
    # if element <= pivot, swap it with the element at current "greaterElemIndex"
    for i in range(left, right):
        if array[i] <= pivot:
            array[i], array[firstGreaterElemIndex] = array[firstGreaterElemIndex], array[i]
            firstGreaterElemIndex += 1
    
    # swap the pivot element with the element at greaterElemIndex
    array[firstGreaterElemIndex], array[right] = array[right], array[firstGreaterElemIndex]
    return firstGreaterElemIndex

def quicksort(array, left, right):
    if left < right:
        pivotIndex = partition(array, left, right)
        # recursive call on the left of pivot (sort the portion which is less than pivot)
        quicksort(array, left, pivotIndex-1)
        # recursive call on the right of pivot (sort the portion which is greater than pivot)
        quicksort(array, pivotIndex+1, right)


if __name__ == "__main__":
    array = [1, 7, 4, 1, 10, 9, -2]
    print("unsorted array: ")
    print(array)

    quicksort(array, 0, len(array)-1)
    print("sorted array: ")
    print(array)
